src/modules.rs

```diff
@@ -4,13 +4,69 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+use std::cell::{Ref, RefCell};
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::compiler::WasmerCompiler;
 use crate::VMError;
 
+use cached::proc_macro::cached;
 use parity_wasm::elements;
+use wasmer::Module;
 use wasmparser::Validator;
 
+use dusk_abi::host::HostModule;
 pub use dusk_abi::{ContractId, ContractState};
 
+type BoxedHostModule = Box<dyn HostModule>;
+
+/// Compiles a module with the specified bytecode or retrieves it from
+pub fn compile_module(bytecode: &[u8]) -> Result<Module, VMError> {
+    get_or_create_module(bytecode.to_vec())
+}
+
+#[cached(size = 2048, time = 86400, result = true, sync_writes = true)]
+fn get_or_create_module(bytecode: Vec<u8>) -> Result<Module, VMError> {
+    let new_module = WasmerCompiler::create_module(bytecode)?;
+    Ok(new_module)
+}
+
+/// Host module store.
+#[derive(Clone, Default)]
+pub struct Modules(Rc<RefCell<HashMap<ContractId, BoxedHostModule>>>);
+
+pub struct ModuleRef<'a> {
+    map_ref: Ref<'a, HashMap<ContractId, BoxedHostModule>>,
+    id: &'a ContractId,
+}
+
+impl<'a> ModuleRef<'a> {
+    pub fn get(&self) -> Option<&BoxedHostModule> {
+        self.map_ref.get(self.id)
+    }
+}
+
+impl Modules {
+    /// Insert a new module into the store.
+    pub fn insert<M>(&mut self, module: M)
+    where
+        M: 'static + HostModule,
+    {
+        self.0
+            .borrow_mut()
+            .insert(module.module_id(), Box::new(module));
+    }
+
+    /// Get a reference to a particular module from the store.
+    pub fn get_module_ref<'a>(&'a self, id: &'a ContractId) -> ModuleRef<'a> {
+        ModuleRef {
+            map_ref: self.0.borrow(),
+            id,
+        }
+    }
+}
```

Comment:
I like this refactoring, before it was mixed up with registration logic

Reply:
I thought I would take this as an opportunity. It's related to these changes only tangentially, but in my opinion it's ok.

I'm  not entirely happy with `ModuleRef` as a concept, so if you find a way to remove it and return return a `Option<&BoxedHostModule>` directly from modules I'm all ears.